rp2040-hal/examples/gpio_irq_example.rs

```diff
@@ -0,0 +1,187 @@
+/// an infinite loop. If there is an LED connected to that pin, it will blink.
+#[entry]
+fn main() -> ! {
+    // Grab our singleton objects
+    let mut pac = pac::Peripherals::take().unwrap();
+
+    // Set up the watchdog driver - needed by the clock setup code
+    let mut watchdog = hal::Watchdog::new(pac.WATCHDOG);
+
+    // Configure the clocks
+    let _clocks = hal::clocks::init_clocks_and_plls(
+        XTAL_FREQ_HZ,
+        pac.XOSC,
+        pac.CLOCKS,
+        pac.PLL_SYS,
+        pac.PLL_USB,
+        &mut pac.RESETS,
+        &mut watchdog,
+    )
+    .ok()
+    .unwrap();
+
+    // The single-cycle I/O block controls our GPIO pins
+    let sio = hal::Sio::new(pac.SIO);
+
+    // Set the pins to their default state
+    let pins = hal::gpio::Pins::new(
+        pac.IO_BANK0,
+        pac.PADS_BANK0,
+        sio.gpio_bank0,
+        &mut pac.RESETS,
+    );
+
+    // Configure GPIO 25 as an output to drive our LED.
+    // we can use into_mode() instead of into_pull_up_input()
+    // since the variable we're pushing it into has that type
+    let led = pins.gpio25.into_mode();
+
+    // Set up the GPIO pin that will be our input
+    let in_pin = pins.gpio26.into_mode();
+
+    // Trigger on the 'falling edge' of the input pin.
+    // This will happen as the button is being pressed
+    in_pin.set_interrupt_enabled(EdgeLow, true);
+
+    // Give away our pins by moving them into the `GLOBAL_PINS` variable.
+    // We won't need to access them in the main thread again
+    cortex_m::interrupt::free(|cs| {
+        GLOBAL_PINS.borrow(cs).replace(Some((led, in_pin)));
+    });
+
+    // Unmask the IO_BANK0 IRQ so that the NVIC interrupt controller
+    // will jump to the interrupt function when the interrupt occurs.
+    // We do this last so that the interrupt can't go off while
+    // it is in the middle of being configured
+    unsafe {
+        pac::NVIC::unmask(pac::Interrupt::IO_IRQ_BANK0);
+    }
+
+    loop {
+        // interrupts handle everything else in this example.
+        // if we wanted low power we could go to sleep. to
+        // keep this example simple we'll just execute a `nop`.
+        // the `nop` (No Operation) instruction does nothing,
+        // but if we have no code here clippy would complain.
+        cortex_m::asm::nop();
+    }
+}
+
+#[interrupt]
+fn IO_IRQ_BANK0() {
+    static mut LED_AND_BUTTON: Option<LedAndButton> = None;
+
+    // This is one-time lazy initialisation. We steal the variables given to us
+    // via `GLOBAL_PINS`.
+    if LED_AND_BUTTON.is_none() {
+        cortex_m::interrupt::free(|cs| {
+            *LED_AND_BUTTON = GLOBAL_PINS.borrow(cs).take();
+        });
+    }
+
+    // Need to check if our Option<LedAndButtonPins> contains our pins
+    if let Some(gpios) = LED_AND_BUTTON {
+        // borrow led and button by *destructuring* the tuple
+        // these will be of type `&mut LedPin` and `&mut ButtonPin`, so we don't have
```

Comment:
It took me about an hour to understand why those types would be mutable references.
Hint: It's because of magic behavior of the `#[interrupt]` attribute, which [converts](https://docs.rs/cortex-m-rt/latest/cortex_m_rt/attr.interrupt.html#properties) `static mut LED_AND_BUTTON: Option<LedAndButton>` to `let LED_AND_BUTTON: &mut Option<LedAndButton>`.

(And it probably would have taken me much longer, if I didn't work on the very similar `#[entry]` macro [during the last few days](https://github.com/rp-rs/rp-hal/issues/292#issuecomment-1038293127).)

Should this be mentioned here to avoid confusion?

Reply:
That's a good idea.
We shouldn't go into too much detail though, don't want to overwhelm beginners with too much info.

Reply:
I've added @jannic's suggested comment to the code to resolve this